module/zfs/zthr.c

```diff
@@ -226,90 +258,150 @@ zthr_create_timer(zthr_checkfunc_t *checkfunc, zthr_func_t *func,
     void *arg, hrtime_t max_sleep)
 {
 	zthr_t *t = kmem_zalloc(sizeof (*t), KM_SLEEP);
-	mutex_init(&t->zthr_lock, NULL, MUTEX_DEFAULT, NULL);
+	mutex_init(&t->zthr_state_lock, NULL, MUTEX_DEFAULT, NULL);
+	mutex_init(&t->zthr_request_lock, NULL, MUTEX_DEFAULT, NULL);
 	cv_init(&t->zthr_cv, NULL, CV_DEFAULT, NULL);
 
-	mutex_enter(&t->zthr_lock);
+	mutex_enter(&t->zthr_state_lock);
 	t->zthr_checkfunc = checkfunc;
 	t->zthr_func = func;
 	t->zthr_arg = arg;
 	t->zthr_wait_time = max_sleep;
 
 	t->zthr_thread = thread_create(NULL, 0, zthr_procedure, t,
 	    0, &p0, TS_RUN, minclsyspri);
-	mutex_exit(&t->zthr_lock);
+	mutex_exit(&t->zthr_state_lock);
 
 	return (t);
 }
 
 void
 zthr_destroy(zthr_t *t)
 {
+	ASSERT(!MUTEX_HELD(&t->zthr_state_lock));
+	ASSERT(!MUTEX_HELD(&t->zthr_request_lock));
 	VERIFY3P(t->zthr_thread, ==, NULL);
-	mutex_destroy(&t->zthr_lock);
+	mutex_destroy(&t->zthr_request_lock);
+	mutex_destroy(&t->zthr_state_lock);
 	cv_destroy(&t->zthr_cv);
 	kmem_free(t, sizeof (*t));
 }
 
 /*
- * Note: If the zthr is not sleeping and misses the wakeup
- * (e.g it is running its ZTHR function), it will check if
- * there is work to do before going to sleep using its checker
- * function [see ZTHR state transition in ZTHR block comment].
- * Thus, missing the wakeup still yields the expected behavior.
+ * Wake up the zthr if it is sleeping. If the thread has been
+ * cancelled that does nothing.
  */
 void
 zthr_wakeup(zthr_t *t)
 {
-	mutex_enter(&t->zthr_lock);
+	mutex_enter(&t->zthr_request_lock);
+	mutex_enter(&t->zthr_state_lock);
+
+	/*
+	 * There are 4 states that we can find the zthr when issuing
+	 * this broadcast:
+	 *
+	 * [1] The common case of the thread being asleep, at which
+	 *     point the broadcast will wake it up.
+	 * [2] The thread has been cancelled and that's probably for
+	 *     a reason (e.g. we are exporting the pool). That's ok,
+	 *     since the checkfunc is the first to run, so next time
+	 *     it is spawned it will pick up the work that we wanted
+	 *     to wake it up for. In this case, this request is a no-op.
```

Comment:
I'm confused about this case. If the thread is cancelled, this function does nothing. Is this trying to say "waking up a cancelled thread is a no-op"?

Reply:
That is right. I just wanted to highlight the fact that the consumer should no worry about this being a no-op. Next time the thread gets spawned, it will look for work before going back to sleep.

Reply:
Ok. I think this point could use a bit of clarification. In particular, I don't really see how the checkfunc is related to any of this.

Reply:
On my previous comment when I say:
`Next time the thread gets spawned, it will look for work before going back to sleep.`
Running the checkfunc is the `look for work` part.

```
static void
zthr_procedure(void *arg)
{
	...
	while (!t->zthr_cancel) {
		if (t->zthr_checkfunc(t->zthr_arg, t)) {
			mutex_exit(&t->zthr_state_lock);
			t->zthr_func(t->zthr_arg, t);
			mutex_enter(&t->zthr_state_lock);
		} else {
			/* go to sleep */
			...
		}
	...
	}
```

Reply:
I think this could be reworded to say:

```
[2] The thread is currently being cancelled. Waking up a cancelled thread is a no-op. Any work that is still left to be done should be handled the next time the thread is resumed.
```